Add file writer and row group writer

sadikovi opened the pull request:
This PR is final part of the core implementation of Parquet writes, adds:
- `FileWriter` and its serialised version `SerializedFileWriter`.
- `RowGroupWriter` and its serialised version `SerializedRowGroupWriter`.

I tried to maintain a fairly simple API. The idea and steps are as follows:
1. Create file writer based on file `std::fs::File`, schema `TypePtr` and writer properties `WriterPropertiesPtr` that you want to write.
2. Call `file_writer.next_row_group` to get a mutable reference to a row group writer to add data.
3. Call `row_group_writer.next_column` to add values to a column. Keep iterating until all columns have been written - we do not support partially written row group!
4. After that you can call `close()` on column writer and/or row group writer. They will be closed regardless, before either the next column is requested or the next row group is requested.
5. Keep adding row groups until you have added enough.
6. Call `file_writer.close` to write final metadata to a file.
7. That is it, after that you should have a complete file, so you can check schema and read records from it.

I added a set of unit tests. Please review them as well, they show API in action.

sunchao commented:
@sadikovi : really sorry that I haven't got chance to look at this yet. Will take a look today.

sadikovi commented:
No worries, take as much time as needed.
On Tue, 21 Aug 2018 at 6:37 PM, Chao Sun <notifications@github.com> wrote:

> @sadikovi <https://github.com/sadikovi> : really sorry that I haven't got
> chance to look at this yet. Will take a look today.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/sunchao/parquet-rs/pull/149#issuecomment-414739703>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHbY3vuyRRb6CI-LUG8TIF2iTgpCybYVks5uTDdegaJpZM4V_Uzh>
> .
>

sunchao commented on src/file/writer.rs:
```diff
@@ -15,16 +15,290 @@
+
+  #[inline]
+  fn next_column(&mut self) -> Result<&mut ColumnWriter> {
+    if self.row_group_metadata.is_some() {
```
nit: we can just call `has_next_column` at the beginning.

sadikovi replied:
I am going to make it iterator-like, this condition will be updated.

sunchao commented on src/file/writer.rs:
```diff
@@ -15,16 +15,290 @@
+  }
+
+  #[inline]
+  fn close(&mut self) -> Result<RowGroupMetaDataPtr> {
```
what if user calls `close()` before iterating over all columns. Will that cause some error?

sadikovi replied:
Yes, it will raise an error, because there will be mismatch between expected columns (must be greater than 0) in the schema and actual number of columns written.

sunchao commented on src/file/writer.rs:
```diff
@@ -15,16 +15,290 @@
+/// File writer interface.
+pub trait FileWriter {
+  /// Creates new row group from this file writer.
+  fn create_row_group(&mut self) -> Result<&mut RowGroupWriter>;
```
The good thing with this is we cannot call `close` until all the returned `RowGroupWriter` are out of scope. The downside is the restrictiveness: now it's hard to pass the `RowGroupWriter` to some struct which lives beyond the current call stack. Not sure if this will be an issue.

sadikovi replied:
Yes, I had the same thought, but it looks like this is one of the few options to keep track of row group writers ourselves. Do you have other ideas? Should we just keep a reference count of row group writers and check later when asking for the next one?

sunchao replied:
Yes, I think another approach is to return a unique reference to the `RowGroupWriter`:
```
fn create_row_group(&mut self) -> Result<RowGroupWriter>;
```

And a separate function to explicitly close it:
```
fn close_row_group(&mut self, RowGroupWriter) -> Result<()>;
```

This will call `finalise_row_group_writer` to close the row group writer and set some flag in the file writer indicating that the current row group writer has been closed.  We can do similar things for `RowGroupWriter` to make sure all column writers are handed back before the `close` is called. 

Let me know what you think. This is just some rough thought.

sadikovi replied:
Sounds good, I will make changes, see what it looks like, and let you know. Thanks!

sadikovi commented on src/file/writer.rs:
```diff
@@ -15,16 +15,290 @@
+  ///
+  /// This method finalises previous column writer, so make sure that all writes are
+  /// finished before calling this method.
+  fn next_column(&mut self) -> Result<&mut ColumnWriter>;
```
I would have to return `Result<Option<&mut ColumnWriter>>`, is it okay?

sunchao replied:
Yes, I think this will be better. We should differentiate the error case and the case where there's no next.

sadikovi replied:
Yes, I changed the code to return `Result<Option<Something>>` so looks like iterator and it is similar to how we read pages.

sadikovi commented on src/file/writer.rs:
```diff
@@ -15,16 +15,290 @@
+use util::io::{FileSink, Position};
+
+/// File writer interface.
+pub trait FileWriter {
```
Yes, will do.

sadikovi replied:
I updated the method docs to emphasise what you mentioned. There is a follow-up sub-task to update all of the documentation related to writes after we have merged the implementation.

sunchao replied:
OK. Sounds good then.

sadikovi commented on src/file/writer.rs:
```diff
@@ -15,16 +15,290 @@
+impl SerializedFileWriter {
+  /// Creates new file writer.
+  pub fn new(mut file: File, schema: TypePtr, properties: WriterPropertiesPtr) -> Self {
+    Self::start_file(&mut file).expect("Start Parquet file");
```
I will update the `new` method to return `Result<Self>` similar to `FileReader` and improve the error message.

sadikovi commented:
@sunchao I addressed your comments, but might have missed some. Would you mind doing another review pass, including tests (I feel they are a bit overcomplicated)? And could you also give your thoughts on the questions I raised in my first comment?

Thanks!

sadikovi commented:
@sunchao I made the changes to the API as per your suggestion. Can you have a look? Thanks!

Now we have `next_XYZ` and `close_XYZ`, we will raise an error if next method is called before close method is called for the previous XYZ.

sunchao commented:
@sadikovi : the latest PR looks good to me in general. However, the test `test_column_writer_check_metadata` is failing. Can you fix that?

sadikovi commented:
Yes, thanks. I just checked the tests, will fix shortly, make some other
minor changes and ping you again.
On Fri, 31 Aug 2018 at 10:46 AM, Chao Sun <notifications@github.com> wrote:

> @sadikovi <https://github.com/sadikovi> : the latest PR looks good to me
> in general. However, the test test_column_writer_check_metadata is
> failing. Can you fix that?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sunchao/parquet-rs/pull/149#issuecomment-417597149>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHbY3uJ5jEm7WWh0bZ7u1q2WaJObELlLks5uWPfXgaJpZM4V_Uzh>
> .
>

sadikovi commented:
@sunchao I fixed the test - it failed because I added test on encodings in this branch, but in master we have updated to add levels encoding as well, so it was failing with missing encoding for levels.

Should be all good now, would appreciate if you could look over the code again. There is also an open question about limiting row groups somehow, either based on size or number of rows - currently we do not limit them in any way. We could also add this as a follow-up issue.

sunchao commented:
> Should be all good now, would appreciate if you could look over the code again. There is also an open question about limiting row groups somehow, either based on size or number of rows - currently we do not limit them in any way. We could also add this as a follow-up issue.

Yes we should have some config properties for this. Good for a follow-up. :)

sunchao commented:
Merged. Thanks @sadikovi !

sadikovi commented:
Thanks!
On Fri, 31 Aug 2018 at 8:48 PM, Chao Sun <notifications@github.com> wrote:

> Merged. Thanks @sadikovi <https://github.com/sadikovi> !
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/sunchao/parquet-rs/pull/149#issuecomment-417757558>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AHbY3s1WzMAaaSJdyY6K9K4a31reIExWks5uWYTtgaJpZM4V_Uzh>
> .
>
